File: algobuddy/src/components/python/algos/sort/merge-sort.py
import json

def log_info(info):
    """
    Log the information in JSON format with sorted keys.
    In a real application, this could write to a file,
    send messages to a visualization module, or update a GUI component.
    """
    print(json.dumps(info, sort_keys=True))

def merge_sort(arr):
    """
    Perform merge sort on an array while logging each step.

    Args:
        arr (list): A list of numbers.
    
    Returns:
        list: The sorted list.
    """
    # Log the current merge sort call with the subarray.
    log_info({"action": "merge_sort_call", "subarray": arr.copy()})
    
    if len(arr) > 1:
        mid = len(arr) // 2
        left_half = arr[:mid]
        right_half = arr[mid:]
        
        # Log the split event.
        log_info({"action": "split", "left": left_half, "right": right_half})
        
        merge_sort(left_half)
        merge_sort(right_half)
        
        i = j = k = 0
        
        # Merge the two halves while logging comparisons and merge steps.
        while i < len(left_half) and j < len(right_half):
            # Log the comparison between elements.
            log_info({
                "action": "compare",
                "left_index": i,
                "left_value": left_half[i],
                "right_index": j,
                "right_value": right_half[j]
            })
            
            if left_half[i] < right_half[j]:
                arr[k] = left_half[i]
                i += 1
            else:
                arr[k] = right_half[j]
                j += 1
            
            k += 1
            # Log the current state of the merge.
            log_info({
                "action": "merge_step",
                "merged_so_far": arr[:k],
                "left_remaining": left_half[i:],
                "right_remaining": right_half[j:]
            })
        
        # Process remaining elements of left_half.
        while i < len(left_half):
            arr[k] = left_half[i]

File: algobuddy/src/components/JavaScript/algos/sort/merge-sort.js
import { logInfo } from '../logger.js';

export function mergeSort(arr) {
  // Log the current merge sort call with the subarray.
  logInfo({ action: "merge_sort_call", subarray: [...arr] });
  
  if (arr.length > 1) {
    const mid = Math.floor(arr.length / 2);
    const leftHalf = arr.slice(0, mid);
    const rightHalf = arr.slice(mid);
    
    // Log the split event.
    logInfo({ action: "split", left: leftHalf, right: rightHalf });
    
    // Recursively sort left and right halves.
    mergeSort(leftHalf);
    mergeSort(rightHalf);
    
    let i = 0, j = 0, k = 0;
    
    // Merge the two halves while logging comparisons and merge steps.
    while (i < leftHalf.length && j < rightHalf.length) {
      // Log the comparison between elements.
      logInfo({
        action: "compare",
        left_index: i,
        left_value: leftHalf[i],
        right_index: j,
        right_value: rightHalf[j]
      });
      
      if (leftHalf[i] < rightHalf[j]) {
        arr[k] = leftHalf[i];
        i++;
      } else {
        arr[k] = rightHalf[j];
        j++;
      }
      k++;
      
      // Log the current state of the merge.
      logInfo({
        action: "merge_step",
        merged_so_far: arr.slice(0, k),
        left_remaining: leftHalf.slice(i),
        right_remaining: rightHalf.slice(j)
      });
    }
    
    // Process remaining elements of leftHalf.
    while (i < leftHalf.length) {
      arr[k] = leftHalf[i];
      i++;
      k++;
      logInfo({
        action: "merge_step",
        merged_so_far: arr.slice(0, k),
        left_remaining: leftHalf.slice(i),
        right_remaining: rightHalf.slice(j)
      });
    }
    
    // Process remaining elements of rightHalf.
    while (j < rightHalf.length) {
      arr[k] = rightHalf[j];
      j++;
      k++;
      logInfo({
        action: "merge_step",
        merged_so_far: arr.slice(0, k),
        left_remaining: leftHalf.slice(i),
        right_remaining: rightHalf.slice(j)
      });
    }
    
    // Log the complete merge event for this subarray.
    logInfo({ action: "merge_complete", merged_result: [...arr] });
  } else {
    // Log base case where the subarray is of length 1.
    logInfo({ action: "base_case", subarray: [...arr] });
  }
  
  return arr;
}

// Example usage (Driver Code):
if (typeof window === "undefined") {
  let arr = [64, 34, 25, 12, 22, 11, 90];
  mergeSort(arr);
  console.log("Sorted array is:", arr);
}
